File: mobile-client/android/UseCase/System/Billing/UploadReceiptUseCase.kt
package app.mobiling.client.client.usecase.system.billing

import app.mobiling.client.client.contract.system.billing.ReceiptUploadPayload
import app.mobiling.client.client.data.system.billing.BillingReceiptGateway

class UploadReceiptUseCase(private val billingReceiptGateway: BillingReceiptGateway) {
    operator fun invoke(token: String, product: String): Boolean =
        billingReceiptGateway.upload(ReceiptUploadPayload(token = token, product = product))
}

File: mobile-client/ios/UseCase/System/Billing/UploadReceiptUseCase.swift
import Foundation

public struct UploadReceiptUseCase {
    private let billingReceiptGateway: BillingReceiptGateway

    public init(billingReceiptGateway: BillingReceiptGateway) {
        self.billingReceiptGateway = billingReceiptGateway
    }

    public func callAsFunction(token: String, product: String) async throws -> Bool {
        try await billingReceiptGateway.upload(
            payload: ReceiptUploadPayload(token: token, product: product)
        )
    }
}
